### ISSUE TITLE
Build fails with icu 64

### ISSUE THREAD

**Issue by HGuillemet** [cyrusimap/cyrus-imapd#2817]
No way to build version 3.0.10, probably due to an incompatibility with ICU 64.2 (on gentoo, with gcc 8.3.0):
A lot of errors related to template used inside extern "C" blocks. For instance:
```
In file included from /usr/lib/gcc/x86_64-pc-linux-gnu/8.3.0/include/g++-v8/memory:82,
                 from /usr/include/unicode/localpointer.h:45,
                 from /usr/include/unicode/uenum.h:23,
                 from /usr/include/unicode/ucnv.h:53,
                 from ./lib/charset.h:61,
                 from imap/xapian_wrap.cpp:17:
/usr/lib/gcc/x86_64-pc-linux-gnu/8.3.0/include/g++-v8/bits/shared_ptr_atomic.h:270:3: error: template with C linkage
   template<typename _Tp, _Lock_policy _Lp>
   ^~~~~~~~
imap/xapian_wrap.cpp:11:1: note: ‘extern "C"’ linkage started here
 extern "C" {
```

**Comment by elliefm** [comment]
This is weird because, I can trace down the same set of headers on my system and find that localpointer.h does indeed declare a template class, but I don't experience this issue during build.

A complicating factor is that I have both ICU 57.1 (from apt, required for libxml2) and ICU 64.2 (from source) installed, but Cyrus is detecting and building with the 64.2 install.

Checking both sets of headers, localpointer.h has the template in either case, so at a glance this doesn't look like something that changed in ICU.

My gcc is older than yours though, `gcc (Debian 6.3.0-18+deb9u1) 6.3.0 20170516`, so I wonder if the issue is something that only the newer gcc complains about?

**Comment by elliefm** [comment]
Actually, it does also fail for me, so I guess I just hadn't built 3.0 since installing ICU 64.2??

**Comment by elliefm** [comment]
Does this patch help?

```
diff --git a/lib/charset.c b/lib/charset.c
index 6c5eaa3e6..22be3ec14 100644
--- a/lib/charset.c
+++ b/lib/charset.c
@@ -44,6 +44,8 @@
 #include <stdlib.h>
 #include <string.h>
 
+#include "unicode/ucnv.h"
+
 #include "assert.h"
 #include "charset.h"
 #include "xmalloc.h"
diff --git a/lib/charset.h b/lib/charset.h
index 63a45f509..bede384ee 100644
--- a/lib/charset.h
+++ b/lib/charset.h
@@ -58,8 +58,6 @@
 
 #define CHARSET_UNKNOWN_CHARSET (NULL)
 
-#include "unicode/ucnv.h"
-
 #include "util.h"
 
 typedef int comp_pat;
```

**Comment by elliefm** [comment]
The above commit is on master, though it doesn't seem to make any difference there (for me).

**Comment by HGuillemet** [comment]
Compilation of master fails for me with:
```
make[2]: Entering directory '/home/hg/cyrus-imapd'
/bin/sh ./libtool  --tag=CC   --mode=compile gcc -DHAVE_CONFIG_H -I.  -I/home/hg/cyrus-imapd/com_err/et -I. -I./lib -I. -I./lib -DLIBEXEC_DIR=\"/usr/local/libexec\" -DSBIN_DIR=\"/usr/local/sbin\" -DSYSCONF_DIR=\"/etc\" -DHAVE_CONFIG_H           -fPIC  -fvisibility=hidden -g -O2 -MT imap/libcyrus_imap_la-http_client.lo -MD -MP -MF imap/.deps/libcyrus_imap_la-http_client.Tpo -c -o imap/libcyrus_imap_la-http_client.lo `test -f 'imap/http_client.c' || echo './'`imap/http_client.c
libtool: compile:  gcc -DHAVE_CONFIG_H -I. -I/home/hg/cyrus-imapd/com_err/et -I. -I./lib -I. -I./lib -DLIBEXEC_DIR=\"/usr/local/libexec\" -DSBIN_DIR=\"/usr/local/sbin\" -DSYSCONF_DIR=\"/etc\" -DHAVE_CONFIG_H -fPIC -fvisibility=hidden -g -O2 -MT imap/libcyrus_imap_la-http_client.lo -MD -MP -MF imap/.deps/libcyrus_imap_la-http_client.Tpo -c imap/http_client.c  -fPIC -DPIC -o imap/.libs/libcyrus_imap_la-http_client.o
imap/http_client.c:57:10: fatal error: imap/http_err.h: No such file or directory
```

Concerning ICU, the problem may well be related to: [https://github.com/unicode-org/icu/pull/613](url).

**Comment by elliefm** [comment]
`imap/http_err.h` is a generated file, so something has gone quite weird with your build if it gets as far as compiling `imap/http_client.c` doesn't already have it.  Do you mind raising a separate issue for that please? Your config.log would be useful to see.

I believe the patch I pasted above should fix the 3.0/64.2 issue. It fixes it for me, but I'd appreciate if you could try it out and confirm too, because my system is a little weird at the moment and that might be masking something else.

**Comment by HGuillemet** [comment]
Separated issue created.
To work around it, I applied your patch above to 3.0.10 and it does fix the ICU-related issue.
Thank you !

**Comment by elliefm** [comment]
Thanks heaps for verifying that!  The ICU fix is now also on the cyrus-imapd-3.0 branch and will be included in future 3.0.x releases :)
